tofhir-server/src/main/scala/io/tofhir/server/service/fhir/FileSystemTreeStructureService.scala

```diff
@@ -25,20 +26,42 @@ class FileSystemTreeStructureService {
    * @return
    */
   def getFolderTreeStructure(basePath: String, includeFiles: Boolean): FilePathNode = {
+    var fileCount = 0 // Track total number of files processed
+
     def buildNode(path: Path): FilePathNode = {
+      if (fileCount >= FILE_LIMIT) {
+        throw new IllegalStateException(s"File limit exceeded. The maximum number of files to process is $FILE_LIMIT.")
+      }
+
       val children = if (Files.isDirectory(path)) {
-        Files.list(path).iterator().asScala
-          .filter(p => includeFiles || Files.isDirectory(p))
-          .filterNot(shouldBeExcluded) // Filter out hidden and specific folders
-          .map(buildNode)
-          .toList
+        val dirStream = Files.list(path)
+        try {
+          dirStream.iterator().asScala
+            .filter(p => includeFiles || Files.isDirectory(p))
+            .filterNot(shouldBeExcluded)  // Filter out hidden files, folders
+            .map { p =>
+              fileCount += 1
+              buildNode(p)
+            }
+            .toList
+        } finally {
+          dirStream.close() // Ensures the stream is closed after usage
+        }
       } else {
         List.empty[FilePathNode]
       }
       FilePathNode(path.getFileName.toString, Files.isDirectory(path), children)
     }
+
     val path = FileUtils.getPath(basePath)
+    if(!path.toAbsolutePath.normalize().startsWith(FileUtils.getPath("").toAbsolutePath)) {
+      throw new IllegalArgumentException("The given path is outside the root context path.")
+    }
     buildNode(path)
   }
 
+}
+
+object FileSystemTreeStructureService {
+  val FILE_LIMIT = 10000 // Maximum number of files to process
```

Comment:
Should we make it configurable via `application.conf` ?

Reply:
Sure, but I'm not sure where should we put this config. It is a tofhir-server specific config and we currently don't have a config for tofhir-server specifically. There is web-server config but it's more related to http stuff. For now, let's not define a separate config file just for this. It would be better if it was configurable, but there is no need for it to be configurable at the moment. If we need tohir-server specific configs in the future, we can add that one as well.